mat_dp_pipeline/pipeline/common.py

```diff
@@ -0,0 +1,135 @@
+from dataclasses import dataclass
+from pathlib import Path
+
+import pandas as pd
+
+import mat_dp_pipeline.sdf.standard_data_format as sdf
+from mat_dp_pipeline.common import validate
+
+
+@dataclass(eq=False, order=False)
+class SparseYearsInput:
+    """Data amalgamated from hierachical structure, with potential gaps between the years.
+
+    The gaps between years are later filled with values (interpolated), but this isn't in scope
+    of this class.
+
+    Year is another dimension here -- level 0 index in intensities & indicators,
+    year columns in targets.
+
+    Attributes:
+        intensities (DataFrame): (Year, Tech) x (Resource1, Resource2, ..., ResourceN)
+        targets (DataFrame): Tech x  (Year1, Year2, ..., YearK). Tech keys must be a subset of
+            intensities' keys
+        indicators (DataFrame): (Year, Resource) x (Indicator1, Indicator2, ..., IndicatorM)
+            There should be exactly N resources, matching columns in intensities frame
+        tech_metadata (DataFrame): Technologies metadata
+    """
+
+    intensities: pd.DataFrame
+    targets: pd.DataFrame
+    indicators: pd.DataFrame
+    tech_metadata: pd.DataFrame
+
+    def copy(self) -> "SparseYearsInput":
+        return SparseYearsInput(
+            intensities=self.intensities.copy(),
+            targets=self.targets.copy(),
+            indicators=self.indicators.copy(),
+            tech_metadata=self.tech_metadata.copy(),
+        )
+
+    def validate(self) -> set[str]:
+        """Validates whether an object represents a valid instance of
+        CombinedInput. Apart from validation, it also narrows down the set of
+        techs and resources to match the frames.
+
+        Raises:
+            ValueError: when validation fails
+
+        Returns:
+            set[str]: mismatched resources. It's up to the client code to decide
+            whether it's an error or not. No exception is thrown here if this set
+            isn't empty.
+        """
+        sdf.validate_tech_units(self.tech_metadata)
+        intensities_techs = self.intensities.index.droplevel(0).unique()
+        targets_techs = self.targets.index
+
+        validate(
+            set(targets_techs) <= set(intensities_techs),
+            "Target's techs must be a subset of intensities' techs!",
+        )
+
+        # move years to columns, reindex techs, bring the years back
+        self.intensities = (
+            self.intensities.unstack(level=0)  # type: ignore
```

Comment:
type ignore

Reply:
I'm not sure what do you want me to do with all the comments about type: ignore. It is what it is, it doesn't work. Unless you have a solution to this.

Reply:
Can review - was tagging them so they're easier to locate